### ISSUE TITLE
[FEATURE REQUEST] Move some features of Relm4 to gtk-rs

### ISSUE THREAD

**Issue by AaronErhardt** [gtk-rs/gtk4-rs#1038]
Relm4 and gtk-rs share the goal to make GUI development with GTK on Rust as convenient as possible. Over time, Relm4 has gained features that might be useful for gtk-rs users too. Therefore, I thought it would be nice to write down a list of features that don't depend on Relm4 and could be moved to gtk-rs.

Of course, this is just a suggestion and you can decide which features are suitable to be included and how they should be included.

### Features

+ Type-safe actions: [example](https://github.com/Relm4/Relm4/blob/main/relm4-examples/examples/actions.rs#L77-L105) and [implementation](https://github.com/Relm4/Relm4/blob/next/relm4/src/actions/mod.rs)
+ Menu macro to create menus conveniently, together with type-safe actions: [example](https://github.com/Relm4/Relm4/blob/main/relm4-examples/examples/menu.rs#L72-L87)
+ Multithreaded async runtime based on glib (not published yet)
+ `view!` macro to initialize widgets conveniently (even without the relm4 crate)

### Extension traits

+ [WidgetPlus](https://github.com/Relm4/Relm4/blob/next/relm4/src/util/widget_plus.rs): convenience methods that extend WidgetExt
+ [RelmIterChildrenExt](https://github.com/Relm4/Relm4/blob/next/relm4/src/extensions/iter_children.rs): Iterate over the children of widgets
+ [RelmContainerExt](https://github.com/Relm4/Relm4/blob/next/relm4/src/extensions/container.rs): A re-implementation of ContainerExt for suitable widgets
+ [RelmSetChildExt](https://github.com/Relm4/Relm4/blob/next/relm4/src/extensions/set_child.rs): `container_set_child` method
+ [RelmRemoveExt](https://github.com/Relm4/Relm4/blob/next/relm4/src/extensions/remove.rs): `container_remove` method

**Comment by Hofer-Julian** [comment]
> * Type-safe actions: [example](https://github.com/Relm4/Relm4/blob/main/relm4-examples/examples/actions.rs#L77-L105) and [implementation](https://github.com/Relm4/Relm4/blob/next/relm4/src/actions/mod.rs)
> * Menu macro to create menus conveniently, together with type-safe actions: [example](https://github.com/Relm4/Relm4/blob/main/relm4-examples/examples/menu.rs#L72-L87)
> * Multithreaded async runtime based on glib (not published yet)

Especially those sound very useful to me

**Comment by sdroege** [comment]
> Multithreaded async runtime based on glib (not published yet)

I think that should be a separate crate as that gets away from bindings quite a bit. Sounds interesting though. I assume that's juggling lots of main contexts and then schedules tasks between them in some way?

**Comment by AaronErhardt** [comment]
Yes, I planned to create a crate called "glib_async_executor" for this. Indeed, it's just a bunch of main contexts and a [flume channel](https://docs.rs/flume/latest/flume/fn.unbounded.html) in a nutshell. It doesn't have outstanding latencies but works quite well otherwise. I think I will publish my local code very soon.

**Comment by sdroege** [comment]
I have more questions about that but let's discuss that in a separate issue or PR once the code is there :)

**Comment by AaronErhardt** [comment]
I've posted the code of the runtime [here](https://github.com/AaronErhardt/glib_async_executor). It still needs quite a bit of polish but the core features are implemented.

**Comment by sdroege** [comment]
As long as it stays that small I guess we can put it behind a feature flag into the glib crate, but I'm not entirely sure what the use case of this is. How are you using it? :)

Also you probably want to implement the various executor traits from the `futures` crate and provide the various functions that other runtimes have like `block_on`, `spawn_local`, etc.

**Comment by AaronErhardt** [comment]
Relm4 allows you to run background tasks in your components (like commands in Elm), for example for HTTP requests or CPU intensive tasks. So far, this has been handled by tokio, but we want to allow users to choose other runtimes: https://github.com/Relm4/Relm4/issues/114

**Comment by AaronErhardt** [comment]
Apart from of the async runtime discussions, what features are you interested in?

**Comment by sdroege** [comment]
The others are all quite GTK specific so I'm the wrong person to ask. For me they all sound quite interesting though

**Comment by Hofer-Julian** [comment]
Any opinions @bilelmoussaoui?

Or something to ask for feedback in the matrix room?

**Comment by bilelmoussaoui** [comment]
Regarding the extensions traits, all of them except [RelmIterChildrenExt](https://github.com/Relm4/Relm4/blob/next/relm4/src/extensions/iter_children.rs) should be suggested as APIs upstream but most of them were removed in gtk3 due to reasons like the removal of GtkContainer API. For type safe actions, there is already a PR for that in gtk-rs-core, maybe helping finishing that one would be great & see if there are things that could be added to it.

The menu! & view! macros looks interesting though. Do you mind sharing some examples how they work or a link to their source code?

**Comment by Hofer-Julian** [comment]
> For type safe actions, there is already a PR for that in gtk-rs-core, maybe helping finishing that one would be great & see if there are things that could be added to it.

One small note here, if possible I think it would be nice if Relm4 and the bindings would end up with the same API in the end.

**Comment by AaronErhardt** [comment]
> Regarding the extensions traits, all of them except [RelmIterChildrenExt](https://github.com/Relm4/Relm4/blob/next/relm4/src/extensions/iter_children.rs) should be suggested as APIs upstream

We don't have a GTK dev in our team and although I know C and can understand most of GTK's source code, some guidance would be appreciated. Also, this is not just an extension of GTK's widgets, but is implemented for some types from libadwaita and libpanel too.

> The menu! & view! macros looks interesting though. Do you mind sharing some examples how they work or a link to their source code?

Docs with examples including the macro expansion and the source code:
+ view: [docs](https://relm4.org/docs/next/relm4_macros/macro.view.html), [code](https://github.com/Relm4/Relm4/tree/next/relm4-macros/src)
+ menu: [docs](https://relm4.org/docs/next/relm4_macros/macro.menu.html), [code](https://github.com/Relm4/Relm4/tree/next/relm4-macros/src/menu)

The view! macro itself doesn't need any Relm4-specific features, but I should note that it supports additional features in certain contexts (for example combined with the [`#[component]`](https://relm4.org/docs/next/relm4_macros/attr.component.html) attribute macro) that only work together with Relm4. Entirely separating those features into another crate does not make sense because a lot of code is shared, but those features could be put behind a feature flag.

The menu! macro is a lot simpler in comparison and does not depend on anything else but the type safe actions implementation of Relm4. I hope we can work together on a type safe actions implementation. Honestly, I'm not entirely sure which PR in gtk-rs-core you were referring to.

**Comment by bilelmoussaoui** [comment]
> We don't have a GTK dev in our team and although I know C and can understand most of GTK's source code, some guidance would be appreciated. Also, this is not just an extension of GTK's widgets, but is implemented for some types from libadwaita and libpanel too.

Opening an issue with API suggestions and seeing how the upstream gtk developers want to go forward is a good idea. The implementation part is a different story but I definitely don't want the bindings to try and fix the original API. gtkmm did that and it ended up in a very bad shape.

**Comment by bilelmoussaoui** [comment]
> I'm not entirely sure which PR in gtk-rs-core you were referring to.

https://github.com/gtk-rs/gtk-rs-core/pull/67

**Comment by AaronErhardt** [comment]
> https://github.com/gtk-rs/gtk-rs-core/pull/67

Maybe we should discuss this separately, but I think this PR is doing something very different for a different use-case. It seems to target use-cases in combination with UI templates, whereas the approach from Relm4 is more lightweight and less opinionated (but rather targets UIs built with Rust code). Naturally, UI templates that use GTK builder can't transport any type information. However, the implementation in Relm4 is explicitly type safe: you can't accidentally pass an action with the wrong signature nor misspell the action name.
